File: Runtime/python/Phrost/Sprite.py
from typing import Any, Dict, List, Optional

from ChannelPacker import ChannelPacker
from Channels import Channels
from Events import Events


class Sprite:
    def __init__(self, id0: int, id1: int, is_new: bool = True):
        self.id0: int = id0
        self.id1: int = id1
        # readonly in Python is just a convention (prefix with _)
        # but for a direct port, we'll keep them public.

        self.position: Dict[str, float] = {"x": 0.0, "y": 0.0, "z": 0.0}
        self.size: Dict[str, float] = {"width": 1.0, "height": 1.0}
        self.color: Dict[str, int] = {"r": 255, "g": 255, "b": 255, "a": 255}
        self.texture_path: Optional[str] = None
        self.rotate: Dict[str, float] = {"x": 0.0, "y": 0.0, "z": 0.0}
        self.speed: Dict[str, float] = {"x": 0.0, "y": 0.0}
        self.scale: Dict[str, float] = {"x": 1.0, "y": 1.0, "z": 1.0}

        self.source_rect: Optional[Dict[str, float]] = None
        self.texture_id: int = 0

        # Stores which properties have changed since the last event pack.
        self.dirty_flags: Dict[str, bool] = {}
        # Flag to track if this sprite was just created.
        self.is_new: bool = is_new

    def update(self, dt: float) -> None:
        if self.speed["x"] == 0.0 and self.speed["y"] == 0.0:
            return
        self.set_position(
            self.position["x"] + self.speed["x"] * dt,
            self.position["y"] + self.speed["y"] * dt,
            self.position["z"],
        )

    def set_position(
        self, x: float, y: float, z: float, notify_engine: bool = True
    ) -> None:
        if (
            self.position["x"] != x
            or self.position["y"] != y
            or self.position["z"] != z
        ):
            self.position["x"] = x
            self.position["y"] = y
            self.position["z"] = z
            if notify_engine:
                self.dirty_flags["position"] = True

    def set_size(self, width: float, height: float, notify_engine: bool = True) -> None:
        if self.size["width"] != width or self.size["height"] != height:
            self.size["width"] = width
            self.size["height"] = height
            if notify_engine:
                self.dirty_flags["size"] = True

    def set_color(
        self, r: int, g: int, b: int, a: int, notify_engine: bool = True
    ) -> None:
        if (
            self.color["r"] != r
            or self.color["g"] != g
            or self.color["b"] != b
            or self.color["a"] != a
        ):
            self.color["r"] = r
            self.color["g"] = g
            self.color["b"] = b
            self.color["a"] = a
            if notify_engine:
                self.dirty_flags["color"] = True

    def set_texture_path(self, path: str, notify_engine: bool = True) -> None:
        if self.texture_path != path:
            self.texture_path = path
            if notify_engine:
                self.dirty_flags["texture"] = True

    def set_rotate(
        self, x: float, y: float, z: float, notify_engine: bool = True
    ) -> None:
        if self.rotate["x"] != x or self.rotate["y"] != y or self.rotate["z"] != z:
            self.rotate["x"] = x
            self.rotate["y"] = y
            self.rotate["z"] = z
            if notify_engine:
                self.dirty_flags["rotate"] = True

    def set_speed(self, x: float, y: float, notify_engine: bool = True) -> None:
        if self.speed["x"] != x or self.speed["y"] != y:
            self.speed["x"] = x
            self.speed["y"] = y
            if notify_engine:
                self.dirty_flags["speed"] = True

    def set_scale(
        self, x: float, y: float, z: float, notify_engine: bool = True
    ) -> None:
        if self.scale["x"] != x or self.scale["y"] != y or self.scale["z"] != z:
            self.scale["x"] = x
            self.scale["y"] = y
            self.scale["z"] = z
            if notify_engine:
                self.dirty_flags["scale"] = True

    def set_flip(self, is_flipped: bool, notify_engine: bool = True) -> None:
        """
        Sets the horizontal flip state of the sprite by modifying its X scale.
        """
        # Get the current absolute X scale
        current_abs_scale_x = abs(self.scale["x"])

        # Determine the new X scale
        new_scale_x = -current_abs_scale_x if is_flipped else current_abs_scale_x

        # Use the existing set_scale method to apply the change
        # This will automatically handle the dirty flag
        self.set_scale(
            new_scale_x,
            self.scale["y"],
            self.scale["z"],
            notify_engine,
        )

    def set_source_rect(
        self, x: float, y: float, w: float, h: float, notify_engine: bool = True
    ) -> None:
        """Sets the source rectangle for texture mapping."""
        new_rect = {"x": x, "y": y, "w": w, "h": h}
        if self.source_rect != new_rect:
            self.source_rect = new_rect
            if notify_engine:
                self.dirty_flags["source_rect"] = True

    def set_texture_id(self, texture_id: int) -> None:
        self.texture_id = texture_id

    # --- Getters (for reading state) ---
    def get_position(self) -> Dict[str, float]:
        return self.position

    def get_speed(self) -> Dict[str, float]:
        return self.speed

    def get_scale(self) -> Dict[str, float]:
        return self.scale

    def get_rotation(self) -> Dict[str, float]:
        return self.rotate

    def get_color(self) -> Dict[str, int]:
        return self.color

    def get_id(self) -> List[int]:
        return [self.id0, self.id1]

    def get_texture_id(self) -> int:
        return self.texture_id

    def get_source_rect(self) -> Optional[Dict[str, float]]:
        """Gets the source rectangle for texture mapping."""
        return self.source_rect

    def get_initial_add_data(self) -> List[Any]:
        """Generates the data array for the initial SPRITE_ADD event."""
        return [
            self.id0,
            self.id1,
            self.position["x"],
            self.position["y"],
            self.position["z"],
            self.scale["x"],
            self.scale["y"],
            self.scale["z"],
            self.size["width"],
            self.size["height"],
            self.rotate["x"],
            self.rotate["y"],
            self.rotate["z"],
            self.color["r"],
            self.color["g"],
            self.color["b"],
            self.color["a"],
            self.speed["x"],
            self.speed["y"],
        ]

    # --- THIS METHOD IS NOW CORRECTED ---
    def pack_dirty_events(self, packer: ChannelPacker, clear=True) -> None:
        """
        Checks all dirty flags and adds the corresponding events
        to the ChannelPacker.
        """
        # --- All 'packer.add' calls now include the Channel ID ---
        RENDER_CHANNEL = Channels.RENDERER.value

        if self.is_new:
            # Send the full SPRITE_ADD event
            packer.add(RENDER_CHANNEL, Events.SPRITE_ADD, self.get_initial_add_data())

            # Also send the texture load event if a texture was set
            if self.texture_path is not None:
                filename_bytes = self.texture_path.encode("utf-8")
                filename_length = len(filename_bytes)
                packer.add(
                    RENDER_CHANNEL,
                    Events.SPRITE_TEXTURE_LOAD,
                    [
                        self.id0,
                        self.id1,
                        filename_length,
                        filename_bytes,  # Pass bytes
                    ],
                )

            # Also send source rect if it was set during initialization
            if self.source_rect is not None:
                packer.add(
                    RENDER_CHANNEL,
                    Events.SPRITE_SET_SOURCE_RECT,
                    [
                        self.id0,
                        self.id1,
                        self.source_rect["x"],
                        self.source_rect["y"],
                        self.source_rect["w"],
                        self.source_rect["h"],
                    ],
                )

            # Mark as no longer new and clear all other flags
            self.is_new = False
            self.clear_dirty_flags()
            return  # Exit

        # --- REGULAR DIRTY CHECK ---
        if not self.dirty_flags:
            return  # Nothing to do

        if "position" in self.dirty_flags:
            packer.add(
                RENDER_CHANNEL,
                Events.SPRITE_MOVE,
                [
                    self.id0,
                    self.id1,
                    self.position["x"],
                    self.position["y"],
                    self.position["z"],
                ],
            )

        if "scale" in self.dirty_flags:
            packer.add(
                RENDER_CHANNEL,
                Events.SPRITE_SCALE,
                [
                    self.id0,
                    self.id1,
                    self.scale["x"],
                    self.scale["y"],
                    self.scale["z"],
                ],
            )

        if "size" in self.dirty_flags:
            packer.add(
                RENDER_CHANNEL,
                Events.SPRITE_RESIZE,
                [
                    self.id0,
                    self.id1,
                    self.size["width"],
                    self.size["height"],
                ],
            )

        if "rotate" in self.dirty_flags:
            packer.add(
                RENDER_CHANNEL,
                Events.SPRITE_ROTATE,
                [
                    self.id0,
                    self.id1,
                    self.rotate["x"],
                    self.rotate["y"],
                    self.rotate["z"],
                ],
            )

        if "color" in self.dirty_flags:
            packer.add(
                RENDER_CHANNEL,
                Events.SPRITE_COLOR,
                [
                    self.id0,
                    self.id1,
                    self.color["r"],
                    self.color["g"],
                    self.color["b"],
                    self.color["a"],
                ],
            )

        if "speed" in self.dirty_flags:
            packer.add(
                RENDER_CHANNEL,
                Events.SPRITE_SPEED,
                [
                    self.id0,
                    self.id1,
                    self.speed["x"],
                    self.speed["y"],
                ],
            )

        if "texture" in self.dirty_flags:
            filename = self.texture_path or ""
            filename_bytes = filename.encode("utf-8")
            filename_length = len(filename_bytes)
            packer.add(
                RENDER_CHANNEL,
                Events.SPRITE_TEXTURE_LOAD,
                [
                    self.id0,
                    self.id1,
                    filename_length,
                    filename_bytes,  # Pass bytes
                ],
            )

        if "source_rect" in self.dirty_flags:
            if self.source_rect is not None:
                packer.add(
                    RENDER_CHANNEL,
                    Events.SPRITE_SET_SOURCE_RECT,
                    [
                        self.id0,
                        self.id1,
                        self.source_rect["x"],
                        self.source_rect["y"],
                        self.source_rect["w"],
                        self.source_rect["h"],
                    ],
                )

        if clear:
            self.clear_dirty_flags()

    def clear_dirty_flags(self) -> None:
        self.dirty_flags = {}

File: Runtime/javascript/Phrost/Sprite.js
const { Events, Channels } = require("./Events");

/**
 * Sprite Class
 * * Base entity for visual objects.
 * Manages properties like position, scale, rotation, color, and texture.
 */
class Sprite {
  /**
   * @param {number} id0 - Primary ID
   * @param {number} id1 - Secondary ID
   * @param {boolean} [isNew=true] - Whether this is a new sprite
   */
  constructor(id0, id1, isNew = true) {
    this.id0 = id0;
    this.id1 = id1;

    this.position = { x: 0.0, y: 0.0, z: 0.0 };
    this.size = { width: 1.0, height: 1.0 };
    this.color = { r: 255, g: 255, b: 255, a: 255 };
    /** @type {?string} */
    this.texturePath = null;
    this.rotate = { x: 0.0, y: 0.0, z: 0.0 };
    this.speed = { x: 0.0, y: 0.0 };
    this.scale = { x: 1.0, y: 1.0, z: 1.0 };
    this.textureId = 0;

    /** @type {?{x: number, y: number, w: number, h: number}} */
    this.sourceRect = null;

    /** @type {Object.<string, boolean>} */
    this.dirtyFlags = {};

    /** @type {boolean} */
    this.isNew = isNew;
  }

  /**
   * Updates the sprite logic (basic physics movement).
   * * @param {number} dt
   */
  update(dt) {
    if (this.speed.x === 0.0 && this.speed.y === 0.0) {
      return;
    }
    this.setPosition(
      this.position.x + this.speed.x * dt,
      this.position.y + this.speed.y * dt,
      this.position.z,
    );
  }

  /**
   * Sets position.
   * * @param {number} x
   * @param {number} y
   * @param {number} z
   * @param {boolean} [notifyEngine=true]
   */
  setPosition(x, y, z, notifyEngine = true) {
    if (
      this.position.x !== x ||
      this.position.y !== y ||
      this.position.z !== z
    ) {
      this.position.x = x;
      this.position.y = y;
      this.position.z = z;
      if (notifyEngine) {
        this.dirtyFlags["position"] = true;
      }
    }
  }

  /**
   * Sets size dimensions.
   * * @param {number} width
   * @param {number} height
   * @param {boolean} [notifyEngine=true]
   */
  setSize(width, height, notifyEngine = true) {
    if (this.size.width !== width || this.size.height !== height) {
      this.size.width = width;
      this.size.height = height;
      if (notifyEngine) {
        this.dirtyFlags["size"] = true;
      }
    }
  }

  /**
   * Sets color.
   * * @param {number} r
   * @param {number} g
   * @param {number} b
   * @param {number} a
   * @param {boolean} [notifyEngine=true]
   */
  setColor(r, g, b, a, notifyEngine = true) {
    if (
      this.color.r !== r ||
      this.color.g !== g ||
      this.color.b !== b ||
      this.color.a !== a
    ) {
      this.color.r = r;
      this.color.g = g;
      this.color.b = b;
      this.color.a = a;
      if (notifyEngine) {
        this.dirtyFlags["color"] = true;
      }
    }
  }

  /**
   * Sets texture path.
   * * @param {string} path
   * @param {boolean} [notifyEngine=true]
   */
  setTexturePath(path, notifyEngine = true) {
    if (this.texturePath !== path) {
      this.texturePath = path;
      if (notifyEngine) {
        this.dirtyFlags["texture"] = true;
      }
    }
  }

  /**
   * Sets rotation (radians).
   * * @param {number} x
   * @param {number} y
   * @param {number} z
   * @param {boolean} [notifyEngine=true]
   */
  setRotate(x, y, z, notifyEngine = true) {
    if (this.rotate.x !== x || this.rotate.y !== y || this.rotate.z !== z) {
      this.rotate.x = x;
      this.rotate.y = y;
      this.rotate.z = z;
      if (notifyEngine) {
        this.dirtyFlags["rotate"] = true;
      }
    }
  }

  /**
   * Sets velocity speed.
   * * @param {number} x
   * @param {number} y
   * @param {boolean} [notifyEngine=true]
   */
  setSpeed(x, y, notifyEngine = true) {
    if (this.speed.x !== x || this.speed.y !== y) {
      this.speed.x = x;
      this.speed.y = y;
      if (notifyEngine) {
        this.dirtyFlags["speed"] = true;
      }
    }
  }

  /**
   * Sets scale.
   * * @param {number} x
   * @param {number} y
   * @param {number} z
   * @param {boolean} [notifyEngine=true]
   */
  setScale(x, y, z, notifyEngine = true) {
    if (this.scale.x !== x || this.scale.y !== y || this.scale.z !== z) {
      this.scale.x = x;
      this.scale.y = y;
      this.scale.z = z;
      if (notifyEngine) {
        this.dirtyFlags["scale"] = true;
      }
    }
  }

  /**
   * Sets the horizontal flip state by modifying X scale.
   * * @param {boolean} isFlipped
   * @param {boolean} [notifyEngine=true]
   */
  setFlip(isFlipped, notifyEngine = true) {
    const currentAbsScaleX = Math.abs(this.scale.x);
    const newScaleX = isFlipped ? -currentAbsScaleX : currentAbsScaleX;

    this.setScale(newScaleX, this.scale.y, this.scale.z, notifyEngine);
  }

  /**
   * Sets the source rectangle for texture mapping.
   * * @param {number} x
   * @param {number} y
   * @param {number} w
   * @param {number} h
   * @param {boolean} [notifyEngine=true]
   */
  setSourceRect(x, y, w, h, notifyEngine = true) {
    // Simple value check logic
    if (
      !this.sourceRect ||
      this.sourceRect.x !== x ||
      this.sourceRect.y !== y ||
      this.sourceRect.w !== w ||
      this.sourceRect.h !== h
    ) {
      this.sourceRect = { x, y, w, h };
      if (notifyEngine) {
        this.dirtyFlags["source_rect"] = true;
      }
    }
  }

  setTextureId(textureId) {
    this.textureId = textureId;
  }

  // --- Getters ---
  getPosition() {
    return this.position;
  }
  getSpeed() {
    return this.speed;
  }
  getScale() {
    return this.scale;
  }
  getRotation() {
    return this.rotate;
  }
  getColor() {
    return this.color;
  }
  getId() {
    return [this.id0, this.id1];
  }
  getTextureId() {
    return this.textureId;
  }
  getSourceRect() {
    return this.sourceRect;
  }

  /**
   * Generates data array for initial SPRITE_ADD.
   * * @returns {Array<number>}
   */
  getInitialAddData() {
    return [
      this.id0,
      this.id1,
      this.position.x,
      this.position.y,
      this.position.z,
      this.scale.x,
      this.scale.y,
      this.scale.z,
      this.size.width,
      this.size.height,
      this.rotate.x,
      this.rotate.y,
      this.rotate.z,
      this.color.r,
      this.color.g,
      this.color.b,
      this.color.a,
      this.speed.x,
      this.speed.y,
    ];
  }

  /**
   * Packs dirty events.
   * * @param {Object} packer
   * @param {boolean} [clear=true]
   */
  packDirtyEvents(packer, clear = true) {
    if (this.isNew) {
      // Send full SPRITE_ADD
      packer.add(
        Channels.RENDERER,
        Events.SPRITE_ADD,
        this.getInitialAddData(),
      );

      // Send texture load if set
      if (this.texturePath) {
        const filename = this.texturePath;
        const filenameLength = Buffer.byteLength(filename, "utf8");
        packer.add(Channels.RENDERER, Events.SPRITE_TEXTURE_LOAD, [
          this.id0,
          this.id1,
          filenameLength,
          filename,
        ]);
      }

      // Send source rect if set
      if (this.sourceRect) {
        packer.add(Channels.RENDERER, Events.SPRITE_SET_SOURCE_RECT, [
          this.id0,
          this.id1,
          this.sourceRect.x,
          this.sourceRect.y,
          this.sourceRect.w,
          this.sourceRect.h,
        ]);
      }

      this.isNew = false;
      this.clearDirtyFlags();
      return;
    }

    if (Object.keys(this.dirtyFlags).length === 0) {
      return;
    }

    if (this.dirtyFlags["position"]) {
      packer.add(Channels.RENDERER, Events.SPRITE_MOVE, [
        this.id0,
        this.id1,
        this.position.x,
        this.position.y,
        this.position.z,
      ]);
    }

    if (this.dirtyFlags["scale"]) {
      packer.add(Channels.RENDERER, Events.SPRITE_SCALE, [
        this.id0,
        this.id1,
        this.scale.x,
        this.scale.y,
        this.scale.z,
      ]);
    }

    if (this.dirtyFlags["size"]) {
      packer.add(Channels.RENDERER, Events.SPRITE_RESIZE, [
        this.id0,
        this.id1,
        this.size.width,
        this.size.height,
      ]);
    }

    if (this.dirtyFlags["rotate"]) {
      packer.add(Channels.RENDERER, Events.SPRITE_ROTATE, [
        this.id0,
        this.id1,
        this.rotate.x,
        this.rotate.y,
        this.rotate.z,
      ]);
    }

    if (this.dirtyFlags["color"]) {
      packer.add(Channels.RENDERER, Events.SPRITE_COLOR, [
        this.id0,
        this.id1,
        this.color.r,
        this.color.g,
        this.color.b,
        this.color.a,
      ]);
    }

    if (this.dirtyFlags["speed"]) {
      packer.add(Channels.RENDERER, Events.SPRITE_SPEED, [
        this.id0,
        this.id1,
        this.speed.x,
        this.speed.y,
      ]);
    }

    if (this.dirtyFlags["texture"]) {
      const filename = this.texturePath || "";
      const filenameLength = Buffer.byteLength(filename, "utf8");
      packer.add(Channels.RENDERER, Events.SPRITE_TEXTURE_LOAD, [
        this.id0,
        this.id1,
        filenameLength,
        filename,
      ]);
    }

    if (this.dirtyFlags["source_rect"] && this.sourceRect) {
      packer.add(Channels.RENDERER, Events.SPRITE_SET_SOURCE_RECT, [
        this.id0,
        this.id1,
        this.sourceRect.x,
        this.sourceRect.y,
        this.sourceRect.w,
        this.sourceRect.h,
      ]);
    }

    if (clear) {
      this.clearDirtyFlags();
    }
  }

  clearDirtyFlags() {
    this.dirtyFlags = {};
  }
}

module.exports = Sprite;
